Adding quantized::conv2d function for pytorch mobile in c10

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26335 Unify Quantization APIs for add, pool and relu
* #26307 Changes to support int8 weight and fp32 bias in QNNPACK
* #26211 Add support to call unpack for pytorch mobile quantized FC and Conv
* **#26152 Adding quantized::conv2d function for pytorch mobile in c10**

Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17459892](https://our.internmc.facebook.com/intern/diff/D17459892)

supriyar committed:
Adding quantized::conv2d function for pytorch mobile in c10
Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -1322,6 +1323,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
+        result_ref_q = torch.quantize_linear(result_NHWK, scale=Y_scale, zero_point=Y_zp, dtype=torch.quint8)
+
+        np.testing.assert_array_almost_equal(result_ref_q.int_repr().numpy(), Y_q.int_repr().numpy(), decimal=0)
+
```
Please also add a test with qnnpack to : https://github.com/pytorch/pytorch/blob/master/test/test_quantized_models.py

supriyar replied:
I will add in a separate PR as we also need to add support to accept int8 weights in QNNPACK.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -127,7 +133,112 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
+      int64_t groups) {
+    auto& ctx = at::globalContext();
+#ifdef USE_FBGEMM
+    if (ctx.preferredQuantizedEngine() == at::QEngine::FBGEMM) {
```
what is the default for the preferred engine? is it going to be set correctly in the mobile build (based on the compile flags)

supriyar replied:
Default is currently set to FBGEMM. I can set the default for mobile build in a separate PR. Right now I'm setting it using `torch.backends.quantized.engine = torch.qnnpack` in the tests.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -111,12 +113,31 @@ class QConv2dInt8 final : public c10::OperatorKernel {
-        ", expected input (NHWC) ", N, ", ", H, ", ", W, ", ", C,
-        " to have ", (packB->inputChannels() * groups),
-        " channels, but got ", C, " channels instead");
+    TORCH_CHECK(
```
nit: maybe disable clang format for it - it looks unreadable:

```
// clang-format off
TORCH_CHECK(...
// clang-format on

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+      int64_t groups,
+      double output_scale,
+      int64_t output_zero_point) {
+    TORCH_CHECK(
```
extract these checks in to a subfunction and share with fbgemm implementation?

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+      int64_t output_zero_point) {
+    TORCH_CHECK(
+        act.ndimension() == 4,
+        "quanitzed::conv2d (qnnpack): Expected activation tensor to be 4-dimensional");
```
quantized typo

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+        "output channels")
+
+    // Allocate output Tensor and a buffer for QNNPACK to use
+    Tensor output = at::_empty_affine_quantized(
```
we don't support per channel quant. Do you have check about it somewhere?

supriyar replied:
Yes, added a check in the prepack function.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+    return output;
+  }
+#endif
+#if defined(USE_FBGEMM) || defined(USE_PYTORCH_QNNPACK)
```
I think you don't need this #ifdef. - you can just have one function and it'd throw if neither of them is available

If it's about warnings that arguments are not used - I think there's a way to disable the warnings.

supriyar replied:
Sounds good. Removed the ifdef.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+#endif
+    TORCH_INTERNAL_ASSERT(
+        "Didn't find engine for operation quantized::conv ",
+        toString(ctx.preferredQuantizedEngine()));
```
would be nice to have a constant declared somewhere for which engines are built in and print it here

supriyar replied:
Yes, we can add that once we get the basic building blocks done :)

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -127,7 +133,112 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
+    return cpp_custom_type_hack::create(std::move(wt_ptr), weight.options());
+  }
+#endif // USE_PYTORCH_QNNPACK
+#if defined(USE_FBGEMM) || defined(USE_PYTORCH_QNNPACK)
```
same comment

dzhulgakov commented on test/test_quantized.py:
```diff
@@ -1322,6 +1323,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
+            Y_scale,
+            Y_zp,
+            use_relu):
+        torch.backends.quantized.engine = torch.qnnpack
```
since it's a global flag, you probably want to have a context manager to unset it. Otherwise it'd mess up other tests

supriyar replied:
Do you have an example of how this can be done?

dzhulgakov replied:
Something like this:
https://github.com/pytorch/pytorch/blob/21ba320cd55bc6eecc280e18c1d8d1212a9737f0/test/jit_utils.py#L509-L515

In general, if the tests would be valid for both fbgemm and qnnpack, we can also make the entire Test class to set the engine in the beginning and instantiate it twice for each engine (potentially with some skips if some tests aren't supported)

dzhulgakov commented on test/test_quantized.py:
```diff
@@ -1322,6 +1323,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
+           Y_scale=st.floats(4.2, 5.6),
+           Y_zp=st.integers(0, 4),
+           use_relu=st.booleans())
+    def test_qconv_qnnpack(
```
shouldn't this test be guarded by qnnpack being present? (we might want to have a function for it)

supriyar replied:
There are some checks at the top of the class which skip if QNNPACK has not been built.

dzhulgakov reviewed (changes requested):
looking good, some comments

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -244,23 +259,147 @@ class QConv2dInt8 final : public c10::OperatorKernel {
-        "This PyTorch installation was not built "
-        "with FBGEMM operators");
+        (outShape[3] == out_ch),
+        "quanitzed::conv2d (qnnpack): Number of filters must be equal to number of "
```
quantized typo in many places

supriyar replied:
I think I was typing in my sleep :)

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422828](https://our.internmc.facebook.com/intern/diff/D17422828)

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422828](https://our.internmc.facebook.com/intern/diff/D17422828)

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422828](https://our.internmc.facebook.com/intern/diff/D17422828)

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422828](https://our.internmc.facebook.com/intern/diff/D17422828)

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422828](https://our.internmc.facebook.com/intern/diff/D17422828)

[ghstack-poisoned]

dzhulgakov commented on test/test_quantized.py:
```diff
@@ -1323,6 +1322,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
             np.testing.assert_equal(
                 Y_q_ref2.int_repr().numpy(), Y_q.int_repr().numpy())
 
+    @given(batch_size=st.integers(1, 3),
```
why do we need separate tests for qnnpack? can't we just use the same set of tests as for fbgemm? (and iterate over available engines in the beginning)

same applies to linear tests above

raghuramank100 replied:
Ensuring numeric consistency of QNNPACK and FBGEMM is also a requirement for quantization aware training to work, so it will be good to check that both kernels pass the same tests. For QNNPACK we should also add tests that allow for full use of dynamic range of wts and activations

supriyar replied:
Discussed offline. Will merge tests in a different PR.

dzhulgakov commented on test/test_quantized.py:
```diff
@@ -1323,6 +1322,129 @@ def test_qlinear_qnnpack(self, batch_size, input_channels, output_channels, use_
+            Y_scale,
+            Y_zp,
+            use_relu):
+        with enable_mobile_quantized_engine():
```
nit: I'd rather rename it to a function taking qengine as an argument, but since it's a test - it's minor

dzhulgakov reviewed (approved):
I guess one of us would need to rebase depending on whether permute PR lands first or not

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -125,19 +131,117 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
+    auto weight_contig = weight.contiguous();
+    auto bias_contig = bias.contiguous();
+    auto wt_ptr =
+        guts::make_unique<PackedConvWeightsQnnp>(PackedConvWeightsQnnp{
```
Where is the weight converted from int8 to uint8?

supriyar replied:
That is in a separate PR -https://github.com/pytorch/pytorch/pull/26307

raghuramank100 reviewed (changes requested):
Few comments/questions. Thanks!

supriyar committed:
Update on "Adding quantized::conv2d function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Conv2d operators

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_qnnpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17459892](https://our.internmc.facebook.com/intern/diff/D17459892)

[ghstack-poisoned]
